### DISCUSSION TITLE
Telemetry consumer API

### DISCUSSION THREAD

**Discussion by jschloemer** [nasa/openmct#6319] [Community Q&A]
It looked like that somewhere on one of the pages, a draft version of a telemetry consumer API was advertised as being in draft. How could I go about trying out that functionality?

Thanks!

**Comment by jvigliotta** [comment]
Hi @jschloemer is this what you're looking for? https://github.com/nasa/openmct/blob/master/API.md#telemetry-api or are you referring to a view that would display the Telemetry? You can also take a look at our [Open MCT Tutorials](https://github.com/nasa/openmct-tutorial).

**Reply by jschloemer** [reply]
In that area, I’m referring to the section that says:

The APIs for requesting telemetry from Open MCT -- e.g. for use in custom views -- are currently in draft state and are being revised. If you'd like to experiment with them before they are finalized, please contact the team via the contact-us link on our website.

Does that help?

**Reply by jvigliotta** [reply]
Ah yes, that's been on the backlog for a while now, keeps getting pushed down for higher priority items. But, in the meantime, if you were to look at some of our views, you can see how they work with the Telemetry API. 

For example, here in our LAD Table Row component, we're requesting telemetry via the `requestCollection` method of our TelemetryAPI: https://github.com/nasa/openmct/blob/ceffee9f2274774ac81f439c9c8b9b2158b9a123/src/plugins/LADTable/components/LADRow.vue#L151-L158

So, a "Telemetry Collection" is a collection of both a historical request and a realtime subscription where the data received is sorted by time and when telemetry is received, it is emitted in an `add` event. We also have `remove` and `clear` events. These events let your view know the current state of the Telemetry data for endpoint which you've requested/subscribed.

In the above example, we're only reacting to the `add` and `clear` events because the strategy for our LAD Tables is 'latest' so we're only dealing with one datum at a time. But other views in Open MCT like the Telemetry Table make use of most of the events that a Telemetry Collection emits, ex: https://github.com/nasa/openmct/blob/ceffee9f2274774ac81f439c9c8b9b2158b9a123/src/plugins/telemetryTable/TelemetryTable.js#L154-L159

There you can see we act on the `remove` event as well as `requestStarted` and `requestEnded`, those latter two are used for indicating a "loading" state to the user of the view.

As I said above, the TelemetryAPI `requestCollection` method combines both historical requests and realtime subscriptions. Previously we were only using the `request` and `subscribe` methods of our TelemetryAPI and you are still able to utilize those if you only need historical or realtime telemetry for your views.

Here is an example of JUST the TelemetryAPI `request`  method being used in our Scatter Plots: https://github.com/nasa/openmct/blob/ceffee9f2274774ac81f439c9c8b9b2158b9a123/src/plugins/charts/scatter/ScatterPlotView.vue#L200

And here is an example of our TelemetryAPI `subscribe` method being used in our Scatter Plots: https://github.com/nasa/openmct/blob/ceffee9f2274774ac81f439c9c8b9b2158b9a123/src/plugins/charts/scatter/ScatterPlotView.vue#L320-L322

We will most likely convert our Scatter Plots to use Telemetry Collections in the future, but in the meantime, you can take a look at it as an example of how to use them individually.

Hopefully this clears things up a little bit about how we use the TelemetryAPI in our views. Once we have more documentation in the API I can update this answer. Feel free to look through other plugins in our repo to see the ways these TelemetryAPI methods are used.
